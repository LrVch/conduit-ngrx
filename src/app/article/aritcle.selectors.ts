import { createSelector, createFeatureSelector } from '@ngrx/store';
import { AppState } from '../reducers';
import { ArticleState } from './article.reducer';

export const selectArticleState = createFeatureSelector<AppState, ArticleState>('article');

export const selectArticle = createSelector(
  selectArticleState,
  (articleState: ArticleState) => articleState.selected
);

export const selectArticleDeleting = createSelector(
  selectArticleState,
  (articleState: ArticleState) => articleState.isArticleDeleting
);

export const selectArticleErrors = createSelector(
  selectArticleState,
  (articleState: ArticleState) => articleState.errors
);

export const selectComments = createSelector(
  selectArticleState,
  (articleState: ArticleState) => articleState.comments
);

export const selectGetCommentsErrors = createSelector(
  selectArticleState,
  (articleState: ArticleState) => articleState.getCommentsErrors
);

export const selectCommentLoading = createSelector(
  selectArticleState,
  (articleState: ArticleState) => articleState.isCommentsLoading
);

export const selectCommentSaving = createSelector(
  selectArticleState,
  (articleState: ArticleState) => articleState.isCommentSaving
);

export const selectAddCommentErrors = createSelector(
  selectArticleState,
  (articleState: ArticleState) => articleState.addCommentErrors
);

export const selecCommentReset = createSelector(
  selectArticleState,
  (articleState: ArticleState) => articleState.isCommentReset
);

export const selectDeleteCommentErrors = createSelector(
  selectArticleState,
  (articleState: ArticleState) => articleState.deleteCommentErrors
);
