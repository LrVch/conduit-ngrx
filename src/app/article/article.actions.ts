import { Action } from '@ngrx/store';
import { Article, Errors, Profile, Comment } from '@app/core';

export enum ArticleActionTypes {
  ArticleLoadSuccess = '[Article] Load Success',
  ArticleClear = '[Article] Load Clear',

  ArticleDeleteRequest = '[Article] Load Delete Request',
  ArticleDeleteSuccess = '[Article] Load Delete Success',
  ArticleDeleteFail = '[Article] Load Delete Fail',

  ArticleToggleFollowingRequest = '[Article] Article Toggle Following Request',
  ArticleToggleFollowingSuccess = '[Article] Article Toggle Following Success',
  ArticleToggleFollowingFail = '[Article] Article Toggle Following Fail',
  ArticleFollowingRequest = '[Article] Article Following Request',
  ArticleUnFollowingRequest = '[Article] Article UnFollowing Request',
  SetFollowingProfile = '[Article] Set Following Profile',
  ClearFollowingProfile = '[Article] Clear Following Profile',

  ArticleCommentsRequest = '[Article] Load Comments Request',
  ArticleCommentsSuccess = '[Article] Load Comments Success',
  ArticleCommentsFail = '[Article] Load Comments Fail',

  ArticleCommentAddRequest = '[Article] Comment Add Request',
  ArticleCommentAddSuccess = '[Article] Comment Add Success',
  ArticleCommentAddFail = '[Article] Comment Add Fail',

  ArticleCommentDeleteRequest = '[Article] Comment Delete Request',
  ArticleCommentDeleteSuccess = '[Article] Comment Delete Success',
  ArticleCommentDeleteFail = '[Article] Comment Delete Fail',
  ArticleDeleteConfirmationRequest = '[Article] Article Delete Confirmation Request',
  ArticleDeleteConfirmation = '[Article] Article Delete Confirmation'
}

export class ArticleLoadSuccess implements Action {
  readonly type = ArticleActionTypes.ArticleLoadSuccess;

  constructor(public payload: { article: Article }) {}
}

export class ArticleDeleteConfirmationRequest implements Action {
  readonly type = ArticleActionTypes.ArticleDeleteConfirmationRequest;

  constructor(public payload: { article?: Article; question: string }) {}
}

export class ArticleDeleteConfirmation implements Action {
  readonly type = ArticleActionTypes.ArticleDeleteConfirmation;

  constructor(public payload: { article: Article }) {}
}

export class ArticleDeleteRequest implements Action {
  readonly type = ArticleActionTypes.ArticleDeleteRequest;

  constructor(public payload: { article: Article }) {}
}

export class ArticleDeleteSuccess implements Action {
  readonly type = ArticleActionTypes.ArticleDeleteSuccess;

  constructor(public payload?: { article?: Article }) {}
}

export class ArticleDeleteFail implements Action {
  readonly type = ArticleActionTypes.ArticleDeleteFail;

  constructor(public payload: { errors: Errors }) {}
}

export class ArticleToggleFollowingRequest implements Action {
  readonly type = ArticleActionTypes.ArticleToggleFollowingRequest;

  constructor(public payload: { profile: Profile }) {}
}

export class ArticleToggleFollowingSuccess implements Action {
  readonly type = ArticleActionTypes.ArticleToggleFollowingSuccess;

  constructor(
    public payload: { profile: Profile; showNotification?: boolean }
  ) {}
}

export class ArticleToggleFollowingFail implements Action {
  readonly type = ArticleActionTypes.ArticleToggleFollowingFail;

  constructor(
    public payload?: { profile: Profile; showNotification?: boolean }
  ) {}
}

export class ArticleFollowingRequest implements Action {
  readonly type = ArticleActionTypes.ArticleFollowingRequest;

  constructor(public payload: { profile: Profile }) {}
}

export class ArticleUnFollowingRequest implements Action {
  readonly type = ArticleActionTypes.ArticleUnFollowingRequest;

  constructor(public payload: { profile: Profile }) {}
}

export class ArticleCommentsRequest implements Action {
  readonly type = ArticleActionTypes.ArticleCommentsRequest;

  constructor(public payload: { slug: string }) {}
}

export class ArticleCommentsSuccess implements Action {
  readonly type = ArticleActionTypes.ArticleCommentsSuccess;

  constructor(public payload: { comments: Comment[] }) {}
}

export class ArticleCommentsFail implements Action {
  readonly type = ArticleActionTypes.ArticleCommentsFail;

  constructor(public payload: { errors: Errors }) {}
}

export class ArticleCommentAddRequest implements Action {
  readonly type = ArticleActionTypes.ArticleCommentAddRequest;

  constructor(public payload: { comment: string }) {}
}

export class ArticleCommentAddSuccess implements Action {
  readonly type = ArticleActionTypes.ArticleCommentAddSuccess;

  constructor(public payload: { comment: Comment }) {}
}

export class ArticleCommentAddFail implements Action {
  readonly type = ArticleActionTypes.ArticleCommentAddFail;

  constructor(public payload: { errors: Errors }) {}
}

export class ArticleCommentDeleteRequest implements Action {
  readonly type = ArticleActionTypes.ArticleCommentDeleteRequest;

  constructor(public payload: { comment: Comment }) {}
}

export class ArticleCommentDeleteSuccess implements Action {
  readonly type = ArticleActionTypes.ArticleCommentDeleteSuccess;

  constructor(public payload: { id: number }) {}
}

export class ArticleCommentDeleteFail implements Action {
  readonly type = ArticleActionTypes.ArticleCommentDeleteFail;

  constructor(public payload: { errors: Errors; id: number }) {}
}

export class SetFollowingProfile implements Action {
  readonly type = ArticleActionTypes.SetFollowingProfile;

  constructor(public payload: { profile: Profile }) {}
}

export class ClearFollowingProfile implements Action {
  readonly type = ArticleActionTypes.ClearFollowingProfile;
}

export type ArticleActions =
  | ArticleLoadSuccess
  | ArticleDeleteConfirmationRequest
  | ArticleDeleteConfirmation
  | ArticleDeleteRequest
  | ArticleDeleteSuccess
  | ArticleDeleteFail
  | ArticleToggleFollowingRequest
  | ArticleToggleFollowingSuccess
  | ArticleToggleFollowingFail
  | ArticleFollowingRequest
  | ArticleUnFollowingRequest
  | ArticleCommentsRequest
  | ArticleCommentsSuccess
  | ArticleCommentsFail
  | ArticleCommentAddRequest
  | ArticleCommentAddSuccess
  | ArticleCommentAddFail
  | ArticleCommentDeleteRequest
  | ArticleCommentDeleteSuccess
  | ArticleCommentDeleteFail
  | SetFollowingProfile
  | ClearFollowingProfile;
