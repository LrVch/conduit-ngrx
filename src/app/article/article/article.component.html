<div class="article-page">
  <app-article-full *ngIf="(article$ | async)"
                    [articleErrors]="articleErrors$ | async"
                    [isDeletingArticle]="isDeletingArticle$ | async"
                    [canModify]="canModify$ | async"
                    [article]="article$ | async"
                    [locale]="locale$ | async"
                    (favorited)="onFavoriveToggle($event)"
                    (folowing)="onToggleFollowing($event)"
                    (delete)="onDeleteArticle($event)">
  </app-article-full>

  <div class="container">
    <div class="row">
      <div class="col-sm-12">

        <div *appShowAuthed="false">
          <br>
          <button sigin
                  (click)="onSignIn()"
                  mat-stroked-button
                  color="primary">
            {{ 'conduit.article.siginButton' | translate }}
          </button>
          {{ 'conduit.article.authCaptionJoin' | translate }}
          <button sigup
                  (click)="onSignUp()"
                  mat-stroked-button
                  color="primary">
            {{ 'conduit.article.sigupButton' | translate }}
          </button>
          {{ 'conduit.article.authCaption' | translate }}.
        </div>

        <div>
          <app-list-error [errors]="addCommentErrors$ | async"></app-list-error>
          <app-list-error [errors]="deleteCommentErrors$ | async"></app-list-error>
          <app-comments *ngIf="(user$ | async)" [comments]="comments$ | async"
                        [reset]="needReset$ | async"
                        [locale]="locale$ | async"
                        (delete)="onDeleteComment($event)"
                        (submitComment)="onSubmitComment($event)"
                        [isSubmitting]="isCommentSaving$ | async"
                        [user]="user$ | async"
                        [loading]="isCommentLoading$ | async"></app-comments>

          <app-list-error [errors]="getCommentsErrors$ | async"></app-list-error>
        </div>
      </div>
    </div>
  </div>

</div>