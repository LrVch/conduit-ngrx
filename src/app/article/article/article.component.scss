.article-page {
    padding-bottom: 50px;

    @media (max-width: 767px) {
        padding-bottom: 30px;
    }
}