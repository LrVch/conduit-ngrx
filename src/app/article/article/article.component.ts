import { Component, OnInit } from '@angular/core';
import { Article, User, Errors, Profile, Comment } from '@app/core';
import { Observable, of } from 'rxjs';
import { Store, select } from '@ngrx/store';
import { AppState } from '@app/reducers';
import {
  selectArticle,
  selectArticleErrors,
  selectArticleDeleting,
  selectComments,
  selectGetCommentsErrors,
  selectCommentLoading,
  selectCommentSaving,
  selectAddCommentErrors,
  selecCommentReset,
  selectDeleteCommentErrors,
} from '../aritcle.selectors';
import { withLatestFrom, switchMap, filter, tap, take, map } from 'rxjs/operators';
import { selectUser } from '@app/auth/auth.selectors';
import {
  ArticleToggleFollowingRequest,
  ArticleCommentsRequest,
  ArticleCommentAddRequest,
  ArticleCommentDeleteRequest,
  ArticleDeleteConfirmationRequest
} from '../article.actions';
import { ToggleArticleFavoriteRequest } from '@app/articles/articles.actions';
import { LogoutAction } from '@app/auth/auth.actions';
import { selectAppSettingsStateLanguage } from '@app/appSettings/app-settings.selectors';

@Component({
  selector: 'app-article',
  templateUrl: './article.component.html',
  styleUrls: ['./article.component.scss']
})
export class ArticleComponent implements OnInit {
  article$: Observable<Article>;
  canModify$: Observable<boolean>;
  isDeletingArticle$: Observable<boolean>;
  articleErrors$: Observable<Errors>;
  user$: Observable<User>;
  comments$: Observable<Comment[]>;
  getCommentsErrors$: Observable<Errors>;
  addCommentErrors$: Observable<Errors>;
  deleteCommentErrors$: Observable<Errors>;
  isCommentLoading$: Observable<boolean>;
  isCommentSaving$: Observable<boolean>;
  needReset$: Observable<boolean>;
  locale$: Observable<string>;

  constructor(
    private store: Store<AppState>
  ) { }

  ngOnInit() {
    this.article$ = this.store.pipe(select(selectArticle));

    this.canModify$ = this.article$.pipe(
      filter(article => !!article),
      withLatestFrom(this.store.pipe(select(selectUser))),
      switchMap(([article, user]: [Article, User]) => {
        return user ? of(article.author.username === user.username) : of(false);
      })
    );

    this.user$ = this.store.pipe(select(selectUser));

    this.locale$ = this.store.pipe(select(selectAppSettingsStateLanguage));

    this.articleErrors$ = this.store.pipe(select(selectArticleErrors));
    this.isDeletingArticle$ = this.store.pipe(select(selectArticleDeleting));

    this.comments$ = this.store.pipe(select(selectComments));
    this.getCommentsErrors$ = this.store.pipe(select(selectGetCommentsErrors));
    this.addCommentErrors$ = this.store.pipe(select(selectAddCommentErrors));
    this.deleteCommentErrors$ = this.store.pipe(select(selectDeleteCommentErrors));
    this.isCommentLoading$ = this.store.pipe(select(selectCommentLoading));
    this.isCommentSaving$ = this.store.pipe(select(selectCommentSaving));
    this.needReset$ = this.store.pipe(select(selecCommentReset));

    this.article$.pipe(
      filter(article => !!article),
      map(article => article.slug),
      tap(slug => this.store.dispatch(new ArticleCommentsRequest({ slug }))),
      take(1)
    ).subscribe();
  }

  onToggleFollowing(profile: Profile): void {
    this.store.dispatch(new ArticleToggleFollowingRequest({ profile }));
  }

  onFavoriveToggle(article: Article): void {
    this.store.dispatch(new ToggleArticleFavoriteRequest({ article }));
  }

  onDeleteArticle(article: Article): void {
    this.store.dispatch(new ArticleDeleteConfirmationRequest({ article, question: 'conduit.article.deleteQuestion' }));
  }

  onDeleteComment(comment: Comment): void {
    this.store.dispatch(new ArticleCommentDeleteRequest({ comment }));
  }

  onSubmitComment(comment: string): void {
    this.store.dispatch(new ArticleCommentAddRequest({ comment }));
  }

  onSignIn() {
    this.store.dispatch(new LogoutAction());
  }

  onSignUp() {
    this.store.dispatch(new LogoutAction({ path: 'register' }));
  }
}
