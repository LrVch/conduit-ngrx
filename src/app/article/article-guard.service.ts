import { Injectable } from '@angular/core';
import {
  CanActivate,
  ActivatedRouteSnapshot,
  RouterStateSnapshot
} from '@angular/router';
import { Observable, of, throwError } from 'rxjs';
import { Store, select } from '@ngrx/store';
import { AppState } from '@app/reducers';
import { Router } from '@angular/router';
import { tap, take, map, switchMap, catchError } from 'rxjs/operators';
import { selectArticle } from './aritcle.selectors';
import { ArticlesService } from '@app/core';
import { ArticleLoadSuccess } from './article.actions';

@Injectable()
export class ArticleGuard implements CanActivate {
  constructor(
    private store: Store<AppState>,
    private router: Router,
    private articlesService: ArticlesService
  ) {}
  canActivate(route: ActivatedRouteSnapshot): Observable<boolean> {
    const slug = route.params['slug'];

    if (!slug) {
      this.router.navigate(['/']);
      return of(false);
    }

    return this.store.pipe(
      take(1),
      switchMap(() => this.articlesService.get(slug)),
      map(article => new ArticleLoadSuccess({ article })),
      tap((action: ArticleLoadSuccess) => this.store.dispatch(action)),
      map(article => !!article),
      catchError(error => {
        if (error.status === '404') {
          this.router.navigate(['/']);
        }
        console.error(error);
        return of(false);
      })
    );
  }
}
