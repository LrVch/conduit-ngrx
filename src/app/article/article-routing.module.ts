import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';
import { AuthGuard } from '@app/core';
import { ArticleComponent } from './article/article.component';
import { ArticleGuard } from './article-guard.service';

const routes: Routes = [
  {
    path: '',
    component: ArticleComponent,
    canActivate: [ArticleGuard],
    data: { title: 'conduit.menu.article' }
  },
  {
    path: ':slug',
    component: ArticleComponent,
    canActivate: [ArticleGuard],
    data: { title: 'conduit.menu.article' }
  }
];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule]
})
export class ArticleRoutingModule {}
