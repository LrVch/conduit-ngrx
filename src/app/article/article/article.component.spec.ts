import { async, ComponentFixture } from '@angular/core/testing';
import { ConfigureFn, configureTests, getUser, getProfile, getArticle, getComment } from '@app/lib/testing';
import { ArticleComponent } from './article.component';
import { DebugElement, Component, Output, EventEmitter, Input } from '@angular/core';
import { User, Profile, Article, Comment } from '@app/core';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { By } from '@angular/platform-browser';
import { Store, StoreModule, combineReducers } from '@ngrx/store';
import * as ArticleActions from '../article.actions';
import * as fromRoot from '@app/reducers';
import * as fromArticle from '../article.reducer';
import * as AuthActions from '@app/auth/auth.actions';
import { hot, cold } from 'jasmine-marbles';
import { ShowAuthedDirective } from '@app/shared';
import { of } from 'rxjs';
import { ToggleArticleFavoriteRequest } from '@app/articles/articles.actions';


@Component({
    selector: 'app-article-full',
    template: `
        <div id="follow" (click)="onToggleFollowing()"></div>
        <div id="favorite" (click)="onFavoriveToggle()"></div>
        <div id="deleteArticle" (click)="onDeleteArticle()"></div>`
})
class TestHostArticleFullComponent {
    @Input('articleErrors') articleErrors: any;
    @Input('isDeletingArticle') isDeletingArticle: any;
    @Input('canModify') canModify: any;
    @Input('article') article: any;
    @Output() favorited = new EventEmitter<any>();
    @Output() folowing = new EventEmitter<any>();
    @Output() delete = new EventEmitter<any>();

    onFavoriveToggle() {
        this.favorited.emit(getArticle());
    }

    onToggleFollowing() {
        this.folowing.emit(getProfile());
    }

    onDeleteArticle() {
        this.delete.emit(getArticle());
    }
}

@Component({
    selector: 'app-list-error',
    template: `<div id="tabs" (click)="onSelectedType($event)"></div>`
})
class TestHostListErrorsComponent {
    @Input('errors') errors: any;
}

@Component({
    selector: 'app-comments',
    template: `
        <div id="submit" (click)="onSubmitComment()"></div>
        <div id="deleteComment" (click)="onDeleteComment()"></div>`
})
class TestHostCommentsComponent {
    @Input('comments') comments: any;
    @Input('reset') reset: any;
    @Input('isSubmitting') isSubmitting: any;
    @Input('user') user: any;
    @Input('loading') loading: any;
    @Output() delete = new EventEmitter();
    @Output() submitComment = new EventEmitter();

    onSubmitComment() {
        this.submitComment.emit('comment');
    }

    onDeleteComment() {
        this.delete.emit(getComment(0));
    }
}

describe('ArticleComponent', () => {
    let component: ArticleComponent;
    let fixture: ComponentFixture<ArticleComponent>;
    let de: DebugElement;
    let store: Store<fromRoot.AppState>;
    let profile: Profile;
    let article: Article;
    let user: User;
    let comment: Comment;

    beforeEach(
        async(() => {
            profile = getProfile();
            user = getUser();
            article = getArticle();
            comment = getComment(0);
            const configure: ConfigureFn = testBed => {
                testBed.configureTestingModule({
                    declarations: [
                        ArticleComponent,
                        TestHostArticleFullComponent,
                        TestHostListErrorsComponent,
                        TestHostCommentsComponent,
                        ShowAuthedDirective
                    ],
                    imports: [
                        NoopAnimationsModule,
                        StoreModule.forRoot({
                            ...fromRoot.reducers,
                            feature: combineReducers(
                                fromArticle.articleReducer,
                            ),
                        }),
                    ],
                });
            };

            configureTests(configure).then(testBed => {
                fixture = testBed.createComponent(ArticleComponent);
                component = fixture.componentInstance;
                de = fixture.debugElement;
                store = testBed.get(Store);
                fixture.detectChanges();

                spyOn(store, 'dispatch').and.callThrough();
            });
        })
    );

    it('should create', () => {
        expect(component).toBeTruthy();
    });

    it('should compile', () => {
        (<any>expect(fixture)).toMatchSnapshot();
    });

    it('should article @Input', () => {
        store.dispatch(new ArticleActions.ArticleLoadSuccess({ article }));
        fixture.detectChanges();
        (<any>expect(fixture)).toMatchSnapshot();
    });

    it('should hide buttons for logged in user', () => {
        store.dispatch(new AuthActions.LoginSuccess({ user }));
        fixture.detectChanges();
        (<any>expect(fixture)).toMatchSnapshot();
    });

    it('should select article', () => {
        component.article$.subscribe(res => {
            expect(res).toBeNull();
        });
    });

    it('should select canModify and set it to true', () => {
        user.username = 'user';
        article.author.username = 'user';
        store.dispatch(new AuthActions.LoginSuccess({ user }));
        store.dispatch(new ArticleActions.ArticleLoadSuccess({ article }));
        component.canModify$.subscribe(res => {
            expect(res).toBeTruthy();
        });
    });

    it('should select canModify and set it to false', () => {
        user.username = '';
        store.dispatch(new AuthActions.LoginSuccess({ user }));
        store.dispatch(new ArticleActions.ArticleLoadSuccess({ article }));
        component.canModify$.subscribe(res => {
            expect(res).toBeFalsy();
        });
    });

    it('should select user', () => {
        const expected = cold('a', { a: null });

        component.ngOnInit();

        expect(component.user$).toBeObservable(expected);
    });

    it('should select articleErrors', () => {
        const expected = cold('a', { a: { errors: {} } });

        component.ngOnInit();

        expect(component.articleErrors$).toBeObservable(expected);
    });

    it('should select isDeletingArticle', () => {
        const expected = cold('a', { a: false });

        component.ngOnInit();

        expect(component.isDeletingArticle$).toBeObservable(expected);
    });

    it('should select comments', () => {
        const expected = cold('a', { a: [] });

        component.ngOnInit();

        expect(component.comments$).toBeObservable(expected);
    });

    it('should select getCommentsErrors', () => {
        const expected = cold('a', { a: { errors: {} } });

        component.ngOnInit();

        expect(component.getCommentsErrors$).toBeObservable(expected);
    });

    it('should select addCommentErrors', () => {
        const expected = cold('a', { a: { errors: {} } });

        component.ngOnInit();

        expect(component.addCommentErrors$).toBeObservable(expected);
    });

    it('should select deleteCommentErrors', () => {
        const expected = cold('a', { a: { errors: {} } });

        component.ngOnInit();

        expect(component.deleteCommentErrors$).toBeObservable(expected);
    });

    it('should select isCommentLoading', () => {
        const expected = cold('a', { a: false });

        component.ngOnInit();

        expect(component.isCommentLoading$).toBeObservable(expected);
    });

    it('should select isCommentSaving', () => {
        const expected = cold('a', { a: false });

        component.ngOnInit();

        expect(component.isCommentSaving$).toBeObservable(expected);
    });

    it('should select needReset', () => {
        const expected = cold('a', { a: false });

        component.ngOnInit();

        expect(component.needReset$).toBeObservable(expected);
    });

    it('should dispatch "ArticleCommentsRequest', () => {
        const spy = jest.spyOn(store, 'pipe').mockImplementation(() => of(article));
        component.ngOnInit();

        expect(store.dispatch).toHaveBeenCalledWith(new ArticleActions.ArticleCommentsRequest({ slug: 'slug' }));
    });

    it('should dispatch "ArticleToggleFollowingRequest"', () => {
        store.dispatch(new ArticleActions.ArticleLoadSuccess({ article }));
        fixture.detectChanges();

        const clickDe = de.query(By.css('#follow'));

        clickDe.triggerEventHandler('click', null);

        expect(store.dispatch).toHaveBeenCalledWith(new ArticleActions.ArticleToggleFollowingRequest({ profile }));
    });

    it('should dispatch "ToggleArticleFavoriteRequest"', () => {
        store.dispatch(new ArticleActions.ArticleLoadSuccess({ article }));
        fixture.detectChanges();

        const clickDe = de.query(By.css('#favorite'));

        clickDe.triggerEventHandler('click', null);

        expect(store.dispatch).toHaveBeenCalledWith(new ToggleArticleFavoriteRequest({ article }));
    });

    it('should dispatch "ArticleDeleteConfirmationRequest"', () => {
        store.dispatch(new ArticleActions.ArticleLoadSuccess({ article }));
        fixture.detectChanges();

        const clickDe = de.query(By.css('#deleteArticle'));

        clickDe.triggerEventHandler('click', null);

        expect(store.dispatch)
            .toHaveBeenCalledWith(
                new ArticleActions.ArticleDeleteConfirmationRequest({ article, question: 'Are you sure you want to delete article?' }));
    });

    it('should dispatch "ArticleCommentAddRequest"', () => {
        store.dispatch(new ArticleActions.ArticleLoadSuccess({ article }));
        fixture.detectChanges();

        const clickDe = de.query(By.css('#submit'));

        clickDe.triggerEventHandler('click', null);

        expect(store.dispatch).toHaveBeenCalledWith(new ArticleActions.ArticleCommentAddRequest({ comment: 'comment' }));
    });

    it('should dispatch "ArticleCommentDeleteRequest"', () => {
        store.dispatch(new ArticleActions.ArticleLoadSuccess({ article }));
        fixture.detectChanges();

        const clickDe = de.query(By.css('#deleteComment'));

        clickDe.triggerEventHandler('click', null);

        expect(store.dispatch).toHaveBeenCalledWith(new ArticleActions.ArticleCommentDeleteRequest({ comment }));
    });

    it('should dispatch "LogoutAction"', () => {
        const clickDe = de.query(By.css('[sigin]'));

        clickDe.triggerEventHandler('click', null);

        expect(store.dispatch).toHaveBeenCalledWith(new AuthActions.LogoutAction());
    });

    it('should dispatch "LogoutAction"', () => {
        const clickDe = de.query(By.css('[sigup]'));

        clickDe.triggerEventHandler('click', null);

        expect(store.dispatch).toHaveBeenCalledWith(new AuthActions.LogoutAction({ path: 'register' }));
    });

});
