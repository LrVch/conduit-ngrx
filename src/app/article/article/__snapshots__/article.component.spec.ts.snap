// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleComponent should article @Input 1`] = `
<app-article
  addCommentErrors$={[Function Store]}
  article$={[Function Store]}
  articleErrors$={[Function Store]}
  canModify$={[Function Store]}
  comments$={[Function Store]}
  deleteCommentErrors$={[Function Store]}
  getCommentsErrors$={[Function Store]}
  isCommentLoading$={[Function Store]}
  isCommentSaving$={[Function Store]}
  isDeletingArticle$={[Function Store]}
  needReset$={[Function Store]}
  store={[Function Store]}
  user$={[Function Store]}
>
  <div
    class="article-page"
  >
    
    <app-article-full
      ng-reflect-article="[object Object]"
      ng-reflect-article-errors="[object Object]"
      ng-reflect-can-modify="false"
      ng-reflect-is-deleting-article="false"
    >
      <div
        id="follow"
      />
      <div
        id="favorite"
      />
      <div
        id="deleteArticle"
      />
    </app-article-full>
    <div
      class="container"
    >
      <div
        class="row"
      >
        <div
          class="col-sm-12"
        >
          
          <div>
            <br />
            <button
              color="primary"
              mat-stroked-button=""
              sigin=""
            >
               Sign in 
            </button>
             or 
            <button
              color="primary"
              mat-stroked-button=""
              sigup=""
            >
               Sign up 
            </button>
             to add comments on this article. 
          </div>
          <div>
            <app-list-error
              ng-reflect-errors="[object Object]"
            >
              <div
                id="tabs"
              />
            </app-list-error>
            <app-list-error
              ng-reflect-errors="[object Object]"
            >
              <div
                id="tabs"
              />
            </app-list-error>
            <app-comments
              ng-reflect-comments=""
              ng-reflect-is-submitting="false"
              ng-reflect-loading="true"
              ng-reflect-reset="false"
            >
              <div
                id="submit"
              />
              <div
                id="deleteComment"
              />
            </app-comments>
            <app-list-error
              ng-reflect-errors="[object Object]"
            >
              <div
                id="tabs"
              />
            </app-list-error>
          </div>
        </div>
      </div>
    </div>
  </div>
</app-article>
`;

exports[`ArticleComponent should compile 1`] = `
<app-article
  addCommentErrors$={[Function Store]}
  article$={[Function Store]}
  articleErrors$={[Function Store]}
  canModify$={[Function Store]}
  comments$={[Function Store]}
  deleteCommentErrors$={[Function Store]}
  getCommentsErrors$={[Function Store]}
  isCommentLoading$={[Function Store]}
  isCommentSaving$={[Function Store]}
  isDeletingArticle$={[Function Store]}
  needReset$={[Function Store]}
  store={[Function Store]}
  user$={[Function Store]}
>
  <div
    class="article-page"
  >
    
    <div
      class="container"
    >
      <div
        class="row"
      >
        <div
          class="col-sm-12"
        >
          
          <div>
            <br />
            <button
              color="primary"
              mat-stroked-button=""
              sigin=""
            >
               Sign in 
            </button>
             or 
            <button
              color="primary"
              mat-stroked-button=""
              sigup=""
            >
               Sign up 
            </button>
             to add comments on this article. 
          </div>
          <div>
            <app-list-error
              ng-reflect-errors="[object Object]"
            >
              <div
                id="tabs"
              />
            </app-list-error>
            <app-list-error
              ng-reflect-errors="[object Object]"
            >
              <div
                id="tabs"
              />
            </app-list-error>
            <app-comments
              ng-reflect-comments=""
              ng-reflect-is-submitting="false"
              ng-reflect-loading="false"
              ng-reflect-reset="false"
            >
              <div
                id="submit"
              />
              <div
                id="deleteComment"
              />
            </app-comments>
            <app-list-error
              ng-reflect-errors="[object Object]"
            >
              <div
                id="tabs"
              />
            </app-list-error>
          </div>
        </div>
      </div>
    </div>
  </div>
</app-article>
`;

exports[`ArticleComponent should hide buttons for logged in user 1`] = `
<app-article
  addCommentErrors$={[Function Store]}
  article$={[Function Store]}
  articleErrors$={[Function Store]}
  canModify$={[Function Store]}
  comments$={[Function Store]}
  deleteCommentErrors$={[Function Store]}
  getCommentsErrors$={[Function Store]}
  isCommentLoading$={[Function Store]}
  isCommentSaving$={[Function Store]}
  isDeletingArticle$={[Function Store]}
  needReset$={[Function Store]}
  store={[Function Store]}
  user$={[Function Store]}
>
  <div
    class="article-page"
  >
    
    <div
      class="container"
    >
      <div
        class="row"
      >
        <div
          class="col-sm-12"
        >
          
          <div>
            <app-list-error
              ng-reflect-errors="[object Object]"
            >
              <div
                id="tabs"
              />
            </app-list-error>
            <app-list-error
              ng-reflect-errors="[object Object]"
            >
              <div
                id="tabs"
              />
            </app-list-error>
            <app-comments
              ng-reflect-comments=""
              ng-reflect-is-submitting="false"
              ng-reflect-loading="false"
              ng-reflect-reset="false"
              ng-reflect-user="[object Object]"
            >
              <div
                id="submit"
              />
              <div
                id="deleteComment"
              />
            </app-comments>
            <app-list-error
              ng-reflect-errors="[object Object]"
            >
              <div
                id="tabs"
              />
            </app-list-error>
          </div>
        </div>
      </div>
    </div>
  </div>
</app-article>
`;
